optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/SolverManager.java

```diff
@@ -0,0 +1,159 @@
+ * interacts with {@code SolverManager}.</li>
+ * <li>Listens for best solution changes and publishes new best solutions via {@link SolutionPublisher}.</li>
+ * </ul>
+ */
+@Component
+class SolverManager implements SolverEventListener<VehicleRoutingSolution> {
+
+    private static final Logger logger = LoggerFactory.getLogger(SolverManager.class);
+
+    private final Solver<VehicleRoutingSolution> solver;
+    private final AsyncTaskExecutor executor;
+    private final SolutionPublisher solutionPublisher;
+
+    private Future<VehicleRoutingSolution> solverFuture;
+
+    SolverManager(
+            Solver<VehicleRoutingSolution> solver,
+            AsyncTaskExecutor executor,
+            SolutionPublisher solutionPublisher
+    ) {
+        this.solver = solver;
+        this.executor = executor;
+        this.solutionPublisher = solutionPublisher;
+        this.solver.addEventListener(this);
+    }
+
+    @Override
+    public void bestSolutionChanged(BestSolutionChangedEvent<VehicleRoutingSolution> bestSolutionChangedEvent) {
+        // CAUTION! This runs on the solver thread. Implications:
+        // 1. The method should be as quick as possible to avoid blocking solver unnecessarily.
+        // 2. This place is a potential source of race conditions.
+        if (!bestSolutionChangedEvent.isEveryProblemFactChangeProcessed()) {
+            logger.info("Ignoring a new best solution that has some problem facts missing");
+            return;
+        }
+        // TODO Race condition, if a servlet thread deletes that location in the middle of this method happening
+        //      on the solver thread. Make sure that location is still in the repository.
+        //      Maybe repair the solution OR ignore if it's inconsistent (log a WARNING).
+        solutionPublisher.publishSolution(bestSolutionChangedEvent.getNewBestSolution()); // TODO @Async
+    }
+
+    void startSolver(VehicleRoutingSolution solution) {
+        if (solverFuture != null) {
+            throw new IllegalStateException("Solver start has already been requested");
+        }
+        // TODO move this to @Async method?
+        // TODO use ListenableFuture to react to solve() exceptions immediately?
+        solverFuture = executor.submit((SolvingTask) () -> solver.solve(solution));
+    }
+
+    void stopSolver() {
+        if (solverFuture != null) {
+            // TODO what happens if solver hasn't started yet (solve() is called asynchronously)
+            solver.terminateEarly();
+            // make sure solver has terminated and propagate exceptions
+            try {
+                solverFuture.get();
+                solverFuture = null;
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException("Failed to stop solver", e);
+            } catch (ExecutionException e) {
+                throw new RuntimeException("Failed to stop solver", e);
```

Comment:
The `ExecutionException` is just a wrapper, you can use directly the `e.getCause()`.